src/lib/lint_syntax/lint_syntax.ml

```diff
@@ -0,0 +1,157 @@
+  in
+  match item with
+  | Pexp_ident id ->
+      version_id id
+  | Pexp_apply ({pexp_desc= Pexp_ident id; _}, _) ->
+      version_id id
+  | _ ->
+      false
+
+let is_version_ident = is_ident "version"
+
+let payload_has_item is_item_ident payload =
+  match payload with
+  | PStr structure ->
+      List.exists structure ~f:(fun str ->
+          match str.pstr_desc with
+          | Pstr_eval (expr, _) -> (
+            (* the "ident" can appear as a singleton ident, or in a tuple *)
+            match expr.pexp_desc with
+            | Pexp_ident _ ->
+                is_item_ident expr.pexp_desc
+            | Pexp_apply ({pexp_desc; _}, _) ->
+                is_item_ident pexp_desc
+            | Pexp_tuple items ->
+                List.exists items ~f:(fun item -> is_item_ident item.pexp_desc)
+            | _ ->
+                false )
+          | _ ->
+              false )
+  | _ ->
+      false
+
+let payload_has_version = payload_has_item is_version_ident
+
+let is_bin_io_ident = is_ident "bin_io"
+
+let payload_has_bin_io = payload_has_item is_bin_io_ident
+
+let attribute_has_deriving_version ((name, payload) : attribute) =
+  String.equal name.txt "deriving" && payload_has_version payload
+
+let attributes_have_deriving_version (attrs : attribute list) =
+  List.exists attrs ~f:attribute_has_deriving_version
+
+let type_has_deriving_version type_decl =
+  attributes_have_deriving_version type_decl.ptype_attributes
+
+let is_deriving name = String.equal name "deriving"
+
+let make_deriving_validator ~pred err_msg type_decl =
+  let get_deriving_error (name, payload) =
+    if is_deriving name.txt then
+      let has_bin_io = payload_has_bin_io payload in
+      let has_version = payload_has_version payload in
+      if pred has_bin_io has_version then Some (name.loc, err_msg) else None
+    else None
+  in
+  List.filter_map type_decl.ptype_attributes ~f:get_deriving_error
+
+let validate_no_bin_io_or_version =
+  make_deriving_validator
+    ~pred:(fun has_bin_io has_version -> has_bin_io || has_version)
+    "deriving bin_io and deriving version disallowed for types in functor body"
+
+let validate_bin_io_and_version =
+  make_deriving_validator
+    ~pred:(fun has_bin_io has_version ->
+      not @@ Bool.equal has_bin_io has_version )
+    "Must have both deriving bin_io and deriving version, or neither"
+
+(* traverse AST, collect errors *)
+let check_deriving_usage =
+  object (self)
+    (* bool indicates whether we're in a functor *)
+    inherit [bool * (Location.t * string) list] Ast_traverse.fold as super
+
+    method get_structure_errors str in_functor =
+      let result =
+        List.map str ~f:(fun si -> self#structure_item si (in_functor, []))
+      in
+      List.concat @@ List.map result ~f:snd
+
+    (* in_functor flows down the tree as we examine subtrees, so we need its value
+       _errors flow up the tree, so we can ignore the passed-in value
```

Comment:
I'm not sure that this is true. The way this is written, I think we end up only returning the errors from the last structure item in a file. Would be good to add a test for this, either way.

I think it might simplify things to
* accumulate the functor depth, hooking into the `module_expr` traverser to increment before and decrement after for `Pmod_functor`
  - the current `in_functor` boolean then becomes equivalent to `functor_depth > 0`
* accumulate the errors by appending them to `errors` at each type declaration

I think this also gets rid of the need for `get_structure_errors`, since the `structure` iterator will then accumulate the errors automatically.